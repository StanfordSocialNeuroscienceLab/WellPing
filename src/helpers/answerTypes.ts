import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
