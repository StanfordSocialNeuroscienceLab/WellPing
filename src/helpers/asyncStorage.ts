import { isToday, addDays } from "date-fns";
import { Notifications } from "expo";
import { AsyncStorage } from "react-native";

import { SurveyScreenState } from "../SurveyScreen";
import PingEntity from "../entities/PingEntity";
import {
  isTimeThisWeekAsync,
  getAllStreamNamesAsync,
  getSurveyFileAsync,
} from "./configFiles";
import { logError } from "./debug";
import { StreamName } from "./types";

export async function getASKeyAsync(key: string = ""): Promise<string> {
  const studyId = (await getSurveyFileAsync()).studyInfo.id;
  return `@WELLPING:Study_${studyId}/${key}`; //TODO: ADD ENCODED URL HERE
}

const NOTIFICATION_TIMES_KEY = `NotificationTime`;

export async function storeNotificationTimesAsync(times: Date[]) {
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(NOTIFICATION_TIMES_KEY),
      JSON.stringify(times),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function clearNotificationTimesAsync() {
  try {
    await Notifications.cancelAllScheduledNotificationsAsync();
    await AsyncStorage.removeItem(await getASKeyAsync(NOTIFICATION_TIMES_KEY));
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function getNotificationTimesAsync(): Promise<Date[] | null> {
  try {
    const value = await AsyncStorage.getItem(
      await getASKeyAsync(NOTIFICATION_TIMES_KEY),
    );
    if (value == null) {
      return null;
    }
    const timesString: string[] = JSON.parse(value);
    const times = timesString.map((dateString) => new Date(dateString));
    return times;
  } catch (error) {
    // Error retrieving data
    logError(error);
    return null;
  }
}

/** TYPEPING TABLE (stored number of pings answered for each type) **/
export async function getNumberOfPingsForStreamName(
  streamName: StreamName,
): Promise<number> {
  const countRaw: {
    count: number;
  } = await PingEntity.createQueryBuilder()
    .select("COUNT(id) as count")
    .where("streamName = :streamName", { streamName })
    .getRawOne();
  return countRaw.count;
}

type NumbersOfPingsForAllStreamNames = {
  [stream: string /* actually StreamName */]: number;
};
export async function getNumbersOfPingsForAllStreamNames(): Promise<
  NumbersOfPingsForAllStreamNames
> {
  const pingsGroupByStreamNameRaw: {
    streamName: StreamName;
    count: number;
  }[] = await PingEntity.createQueryBuilder()
    .select("streamName, COUNT(id) as count")
    .groupBy("streamName")
    .getRawMany();

  return pingsGroupByStreamNameRaw.reduce((map, row) => {
    map[row.streamName] = row.count;
    return map;
  }, {} as NumbersOfPingsForAllStreamNames);
}

/** PING TABLE (stored started ping) **/
export async function insertPingAsync({
  notificationTime,
  startTime,
  streamName,
}: {
  notificationTime: Date;
  startTime: Date;
  streamName: StreamName;
}): Promise<PingEntity> {
  const newIndex = (await getNumberOfPingsForStreamName(streamName)) + 1;
  const pingId = `${streamName}${newIndex}`;
  const tzOffset = startTime.getTimezoneOffset();

  const pingEntity = new PingEntity();
  pingEntity.id = pingId;
  pingEntity.notificationTime = notificationTime;
  pingEntity.startTime = startTime;
  pingEntity.streamName = streamName;
  pingEntity.tzOffset = tzOffset;
  await pingEntity.save();

  return pingEntity;
}

export async function addEndTimeToPingAsync(
  pingId: string,
  endTime: Date,
): Promise<PingEntity> {
  const ping = await PingEntity.createQueryBuilder()
    .where("id = :pingId", { pingId })
    .getOne();
  if (ping == null) {
    throw new Error(`pingId ${pingId} not found in getPingsAsync.`);
  }

  ping.endTime = endTime;
  await ping.save();

  return ping;
}

export async function getLatestPingAsync(): Promise<PingEntity | null> {
  const latestPing = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "DESC")
    .take(1)
    .getOne();
  if (!latestPing) {
    return null;
  }
  return latestPing;
}

export async function getPingsAsync(): Promise<PingEntity[]> {
  const pings = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "DESC")
    .getMany();
  return pings;
}

export async function getTodayPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const todayPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (isToday(ping.notificationTime)) {
      todayPings.push(ping);
    }
    if (ping.notificationTime > addDays(new Date(), 1)) {
      break;
    }
  }
  return todayPings;
}

export async function getThisWeekPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const thisWeekPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (await isTimeThisWeekAsync(ping.notificationTime)) {
      thisWeekPings.push(ping);
    }
    if (ping.notificationTime > new Date()) {
      // Stop when the notification time is in the future.
      break;
    }
  }
  return thisWeekPings;
}

/** SURVEYSTATE TABLE (stores ping answers and other states) **/
const PINGS_STATE_PREFIX = `PingsState:`;

export async function storePingStateAsync(
  pingId: string,
  state: SurveyScreenState,
) {
  try {
    await AsyncStorage.setItem(
      `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`,
      JSON.stringify(state),
    );
  } catch (error) {
    // Error saving data
    logError(error);
    throw error;
  }
}

export async function getPingStateAsync(
  pingId: string,
): Promise<SurveyScreenState> {
  try {
    const keyName = `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`;
    const value = await AsyncStorage.getItem(keyName);
    if (value == null) {
      return {
        currentQuestionId: `ERROR: getPingStateAsync is null for ${keyName}`,
        extraMetaData: {},
        nextStack: [],
        currentQuestionAnswers: {},
        lastUploadDate: new Date(0),
      };
    }
    const state: SurveyScreenState = JSON.parse(value);
    if (state.lastUploadDate) {
      state.lastUploadDate = new Date(state.lastUploadDate);
    }
    return state;
  } catch (error) {
    // Error retrieving data
    logError(error);
    throw error;
  }
}

export async function clearPingStateAsync(pingId: string) {
  try {
    await AsyncStorage.removeItem(
      `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`,
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

/** FUTURE PING QUEUE TABLE (used for storing follow-up streams) **/
const FUTURE_PING_QUEUE_KEY = `FuturePingQueue:`;

export type FuturePing = {
  afterDate: Date;
  streamName: StreamName;
};

type FuturePingJSON = {
  afterDate: string;
  streamName: string;
};

export async function initFuturePingQueueAsync() {
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      JSON.stringify([]),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function enqueueToFuturePingQueue(futurePing: FuturePing) {
  const futurePings = await getFuturePingsQueue();
  if (futurePings == null) {
    throw new Error("futurePings == null in enqueueToFuturePingQueue");
  }
  futurePings.push(futurePing);
  //console.warn(`oy ${JSON.stringify(futurePings)}`);
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      JSON.stringify(futurePings),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function dequeueFuturePingIfAny(): Promise<FuturePing | null> {
  const futurePings = await getFuturePingsQueue();
  if (futurePings == null) {
    throw new Error("futurePings == null in dequeueFuturePingIfAny");
  }
  for (const futurePing of futurePings) {
    if (new Date() > futurePing.afterDate) {
      futurePings.shift();
      try {
        await AsyncStorage.setItem(
          await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
          JSON.stringify(futurePings),
        );
      } catch (error) {
        // Error saving data
        logError(error);
        return null;
      }
      return futurePing;
    }
  }
  return null;
}

export async function getFuturePingsQueue(): Promise<FuturePing[]> {
  try {
    const getValueAsync = async () => {
      return await AsyncStorage.getItem(
        await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      );
    };
    let value = await getValueAsync();
    if (value == null) {
      await initFuturePingQueueAsync();
      value = await getValueAsync();
    }

    if (value == null) {
      throw new Error(
        `${await getASKeyAsync(
          FUTURE_PING_QUEUE_KEY,
        )} is still null after initFuturePingQueueAsync()`,
      );
    }

    const futurePingsJSON: FuturePingJSON[] = JSON.parse(value);
    const futurePings: FuturePing[] = futurePingsJSON.map((futurePingJson) => ({
      ...futurePingJson,
      afterDate: new Date(futurePingJson.afterDate),
      streamName: futurePingJson.streamName,
    }));
    return futurePings;
  } catch (error) {
    // Error retrieving data
    logError(error);
    throw error;
  }
}
