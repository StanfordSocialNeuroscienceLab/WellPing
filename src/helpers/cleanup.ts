import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
