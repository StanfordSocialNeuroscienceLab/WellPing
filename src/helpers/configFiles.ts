import { isThisWeek } from "date-fns";

import { SurveyFile, Names, StudyID, StudyInfo, StreamName } from "./types";

export function getSurveyFile(): SurveyFile {
  const survey: SurveyFile = require("../../config/survey.json");
  return survey;
}

// TODO: DECOUPLE FUNCTIONS LIKE THIS
export function getNamesFile(): Names {
  const names: Names = require("../../config/names.json");
  return names;
}

export function getStudyInfo(): StudyInfo {
  return getSurveyFile().studyInfo;
}

export function getAllStreamNames(): StreamName[] {
  return Object.keys(getSurveyFile().meta.startingQuestionIds) as StreamName[];
}

export function getStudyId(): StudyID {
  return getStudyInfo().id;
}

export function isTimeThisWeek(time: Date): boolean {
  return isThisWeek(time, { weekStartsOn: getStudyInfo().weekStartsOn });
}
