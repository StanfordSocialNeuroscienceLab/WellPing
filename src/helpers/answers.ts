import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
