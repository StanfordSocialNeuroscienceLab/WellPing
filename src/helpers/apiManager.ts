import Constants from "expo-constants";
import * as Crypto from "expo-crypto";
import * as Device from "expo-device";
import { Platform } from "react-native";

import { AnswerEntity } from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { getPingsAsync, getAnswersAsync } from "./asyncStorage";
import { getSurveyFileAsync } from "./configFiles";
import { getUserAsync, User, storeUserAsync } from "./user";

export async function getServerUrlAsync(): Promise<string> {
  return (await getSurveyFileAsync()).studyInfo.serverURL;
}

type UploadData = {
  patientId: string;
  pings: PingEntity[];
  answers: AnswerEntity[];
};

// If success, return `null`. Else return error message.
export async function registerUserAsync(user: User): Promise<string | null> {
  if (!user.patientId || !user.password) {
    return "You should enter both User ID and password.";
  }

  const request: { [key: string]: any } = {
    new_password: user.password, // Do not reset password
    phone_number: "1234567890",
    device_os: Device.osName,
    os_version: Device.osVersion,
    product: Device.modelName,
    brand: Device.brand,
    manufacturer: Device.manufacturer,
    model: Device.modelId,
    beiwe_version: "1.6.0",
  };

  try {
    let endpoint = "/register_user";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    await makePostRequestAsync(endpoint, request, {}, user);
  } catch (e) {
    return `Request error: ${e}.`;
  }

  await storeUserAsync(user);
  return null;
}

export async function getAllDataAsync(): Promise<UploadData> {
  const user = await getUserAsync();
  const pings = await getPingsAsync();
  const answers = await getAnswersAsync();

  const data = {
    patientId: user!.patientId,
    pings,
    answers,
  };
  return data;
}

export async function uploadDataAsync() {
  const user = (await getUserAsync())!;
  const data = await getAllDataAsync();

  try {
    let endpoint = "/ssnl_upload";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    const response = await makePostRequestAsync(endpoint, {}, data, user);
    //console.warn(`YAYA ${JSON.stringify(response)}`);
    return { status: "success", response };
  } catch (e) {
    return { status: "error", error: `Request error: ${e}.` };
  }
}

function base64ToBase64URL(input: string): string {
  return input.replace(/\+/g, "-").replace(/\//g, "_");
}

export async function getRequestURLAsync(
  endpoint: string,
  request: { [key: string]: any } = {},
  forUser?: User,
): Promise<string> {
  let user: User | null = forUser || null;
  if (!user) {
    user = await getUserAsync();

    if (user == null) {
      throw new Error("user == null in getRequestURLAsync");
    }
  }

  const passwordHash = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    user.password,
    {
      encoding: Crypto.CryptoEncoding.BASE64,
    },
  );
  request["password"] = base64ToBase64URL(passwordHash);
  request["device_id"] = `${user.patientId}-${Constants.installationId}`;
  request["patient_id"] = user.patientId;

  //console.warn(`request is ${JSON.stringify(request)}`);

  const serverUrl = await getServerUrlAsync();
  const query = Object.keys(request)
    .map((k) => encodeURIComponent(k) + "=" + encodeURIComponent(request[k]))
    .join("&");
  const url = `${serverUrl}${endpoint}?${query}`;
  return url;
}

export async function makePostRequestAsync(
  endpoint: string,
  request: { [key: string]: any },
  body?: { [key: string]: any },
  user?: User,
): Promise<any> {
  const headers = {
    "Beiwe-Api-Version": "2",
    Accept: "application/vnd.beiwe.api.v2, application/json",
  };

  const url = await getRequestURLAsync(endpoint, request, user);

  //console.warn("url is " + url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  if (response.status < 200 || response.status >= 400) {
    console.warn(`ERRORREQUEST: ${JSON.stringify(response)}`);

    if (response.status === 401 || response.status === 403) {
      throw new Error(
        `Verification failed.\n\nRequest: ${JSON.stringify(
          request,
        )}.\n\nResponse: ${JSON.stringify(response)}.`,
      );
    }

    throw new Error(
      `Request not successful.\n\nRequest: ${JSON.stringify(
        request,
      )}.\n\nResponse: ${JSON.stringify(response)}.`,
    );
  }

  //console.warn(response);

  return response.json();
}
