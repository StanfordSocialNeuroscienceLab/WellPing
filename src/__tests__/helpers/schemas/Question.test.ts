import { QuestionType } from "../../../helpers/helpers";
import {
  QuestionSchema,
  SliderQuestionSchema,
  QuestionTypeSchema,
  ChoicesQuestionSchema,
} from "../../../helpers/schemas/Question";

describe("QuestionTypeSchema", () => {
  test("should not change", () => {
    expect(QuestionTypeSchema).toMatchInlineSnapshot(`
      Object {
        "t": "enum",
        "values": Array [
          "Slider",
          "ChoicesWithSingleAnswer",
          "ChoicesWithMultipleAnswers",
          "YesNo",
          "MultipleText",
          "HowLongAgo",
          "Branch",
          "BranchWithRelativeComparison",
        ],
      }
    `);
  });

  test(".enum should be equal to QuestionType", () => {
    expect(QuestionTypeSchema.enum).toStrictEqual(QuestionType);
  });
});

describe("QuestionSchema", () => {
  test("empty", () => {
    expect(() => {
      QuestionSchema.parse({});
    }).toThrowErrorMatchingSnapshot();
  });

  test("incomplete fields", () => {
    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        type: QuestionType.Slider,
        question: "(This is missing `slider` field)",
        defaultValueFromQuestionId: "Feel_Current",
        next: "Stressor",
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("next should not be undefined", () => {
    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        type: QuestionType.Slider,
        question: "(This is missing `next` field)",
        defaultValueFromQuestionId: "Feel_Current",
        slider: ["left", "right"],
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        type: QuestionType.Slider,
        question: "Good question",
        defaultValueFromQuestionId: "Feel_Current",
        slider: ["left", "right"],
        next: null,
      });
    }).not.toThrowError();

    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        type: QuestionType.Slider,
        question: "Good question",
        defaultValueFromQuestionId: "Feel_Current",
        slider: ["left", "right"],
        next: "Next_Question",
      });
    }).not.toThrowError();
  });

  test("invalid type", () => {
    expect(() => {
      QuestionSchema.parse({
        id: "invalid_question",
        type: "OwO",
        question: "Invalid question!",
        next: null,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("missing type", () => {
    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        question: "Missing type question",
        defaultValueFromQuestionId: "Feel_Current",
        slider: ["left", "right"],
        next: null,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("mismatch type", () => {
    expect(() => {
      QuestionSchema.parse({
        id: "Feel_Ideal",
        type: QuestionType.ChoicesWithSingleAnswer,
        question: "Bad type question",
        defaultValueFromQuestionId: "Feel_Current",
        slider: ["left", "right"],
        next: "Next_Question",
      });
    }).toThrowErrorMatchingSnapshot();
  });
});

describe("SliderQuestionSchema", () => {
  test("type", () => {
    const question = {
      id: "Feel_Ideal",
      question: "Slider question",
      next: "Next_Question",
      slider: ["left", "right"],
    };

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        type: QuestionType.Slider,
      });
    }).not.toThrowError();

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        type: QuestionType.HowLongAgo,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("slider tuple", () => {
    const question = {
      id: "Feel_Ideal",
      type: QuestionType.Slider,
      question: "Slider question",
      next: "Next_Question",
    };

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        slider: ["left", "right"],
      });
    }).not.toThrowError();

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        slider: ["left", "right", "top"],
      });
    }).toThrowErrorMatchingSnapshot("three items");

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        slider: ["left"],
      });
    }).toThrowErrorMatchingSnapshot("one item");

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        slider: [],
      });
    }).toThrowErrorMatchingSnapshot("zero item");
  });

  test("default value range 0 - 100", () => {
    const question = {
      id: "Feel_Ideal",
      type: QuestionType.Slider,
      question: "Range question",
      slider: ["left", "right"],
      next: "Next_Question",
    };

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        defaultValue: 50,
      });
    }).not.toThrowError();

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        defaultValue: 0,
      });
    }).not.toThrowError();

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        defaultValue: 100,
      });
    }).not.toThrowError();

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        defaultValue: -1,
      });
    }).toThrowErrorMatchingSnapshot("<0");

    expect(() => {
      SliderQuestionSchema.parse({
        ...question,
        defaultValue: 101,
      });
    }).toThrowErrorMatchingSnapshot(">100");
  });

  describe("defaultValueFromQuestionId", () => {
    const question = {
      id: "Feel_Ideal",
      type: QuestionType.Slider,
      question: "Range question",
      slider: ["left", "right"],
      next: "Next_Question",
    };

    test("can be undefined", () => {
      expect(() => {
        SliderQuestionSchema.parse({
          ...question,
        });
      }).not.toThrowError();
    });

    test("cannot be null", () => {
      expect(() => {
        SliderQuestionSchema.parse({
          ...question,
          defaultValueFromQuestionId: null,
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test("is question ID", () => {
      expect(() => {
        SliderQuestionSchema.parse({
          ...question,
          defaultValueFromQuestionId: "valid_question_id",
        });
      }).not.toThrowError();

      expect(() => {
        SliderQuestionSchema.parse({
          ...question,
          defaultValueFromQuestionId: "INVALID QUESTION-ID",
        });
      }).toThrowErrorMatchingSnapshot("invalid id");

      expect(() => {
        SliderQuestionSchema.parse({
          ...question,
          defaultValueFromQuestionId: "",
        });
      }).toThrowErrorMatchingSnapshot("empty id");
    });
  });
});

describe("ChoicesQuestionSchema", () => {
  test("type", () => {
    const question = {
      id: "Feel_Ideal",
      question: "Choice question",
      next: "Next_Question",
      choices: [
        { key: "hello", value: "Hello" },
        { key: "world", value: "World" },
      ],
    };

    expect(() => {
      ChoicesQuestionSchema.parse({
        ...question,
        type: QuestionType.ChoicesWithSingleAnswer,
      });
    }).not.toThrowError();

    expect(() => {
      ChoicesQuestionSchema.parse({
        ...question,
        type: QuestionType.ChoicesWithMultipleAnswers,
      });
    }).not.toThrowError();

    expect(() => {
      ChoicesQuestionSchema.parse({
        ...question,
        type: QuestionType.HowLongAgo,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  describe("choices", () => {
    const question = {
      id: "Feel_Ideal",
      type: QuestionType.ChoicesWithSingleAnswer,
      question: "Choice question",
      next: "Next_Question",
    };

    test("should not be undefined", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test("should not be empty", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          choices: [],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test("should be ChoiceSchema objects", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          choices: [
            { key: "hello", value: "Hello" },
            { key: "world", value: "World" },
          ],
        });
      }).not.toThrowError();

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          choices: ["hello", "world"],
        });
      }).toThrowErrorMatchingSnapshot("strings");

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          choices: [2, 3, 5],
        });
      }).toThrowErrorMatchingSnapshot("numbers");
    });
  });

  describe("specialCasesStartId", () => {
    const question = {
      id: "Feel_Ideal",
      type: QuestionType.ChoicesWithSingleAnswer,
      question: "Choice question",
      choices: [
        { key: "hello", value: "Hello" },
        { key: "world", value: "World" },
      ],
      next: "Next_Question",
    };

    test("can be undefined", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
        });
      }).not.toThrowError();
    });

    test("should not be null", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: null,
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test("can be empty object", () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {},
        });
      }).not.toThrowError();
    });

    test(`can include only "_pna"`, () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            _pna: "hello_world",
          },
        });
      }).not.toThrowError();
    });

    test(`can include only choices keys`, () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            hello: "hello_world",
          },
        });
      }).not.toThrowError();
    });

    test(`can include both "_pna" and choices keys`, () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            hello: "hello_world",
            _pna: "goodbye",
          },
        });
      }).not.toThrowError();
    });

    test(`cannot include non-choices key`, () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            notakey: "nonono",
          },
        });
      }).toThrowErrorMatchingSnapshot("non-choices key only");

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            notakey: "nonono",
            world: "stillno",
          },
        });
      }).toThrowErrorMatchingSnapshot("non-choices key with choice key");

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            notakey: "nonono",
            _pna: "goodbye",
          },
        });
      }).toThrowErrorMatchingSnapshot(`non-choices key with "_pna"`);
    });

    test(`question id can be null`, () => {
      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            hello: "new_world",
            _pna: null,
          },
        });
      }).not.toThrowError();

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            world: null,
          },
        });
      }).not.toThrowError();

      expect(() => {
        ChoicesQuestionSchema.parse({
          ...question,
          specialCasesStartId: {
            _pna: null,
          },
        });
      }).not.toThrowError();
    });
  });
});
