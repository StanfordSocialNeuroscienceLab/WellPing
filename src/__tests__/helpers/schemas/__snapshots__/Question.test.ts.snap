// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionSchema empty 1`] = `
"4 validation issue(s)

  Issue #0: invalid_type at id
  Required

  Issue #1: invalid_enum_value at type
  Input must be one of these values: Slider, ChoicesWithSingleAnswer, ChoicesWithMultipleAnswers, YesNo, MultipleText, HowLongAgo, Branch, BranchWithRelativeComparison

  Issue #2: invalid_type at question
  Required

  Issue #3: invalid_type at next
  Required
"
`;

exports[`QuestionSchema incomplete fields 1`] = `
"1 validation issue(s)

  Issue #0: invalid_type at slider
  Required
"
`;

exports[`QuestionSchema invalid type 1`] = `
"1 validation issue(s)

  Issue #0: invalid_enum_value at type
  Input must be one of these values: Slider, ChoicesWithSingleAnswer, ChoicesWithMultipleAnswers, YesNo, MultipleText, HowLongAgo, Branch, BranchWithRelativeComparison
"
`;

exports[`QuestionSchema mismatch type 1`] = `
"2 validation issue(s)

  Issue #0: unrecognized_keys at 
  Unrecognized key(s) in object: 'defaultValueFromQuestionId', 'slider'

  Issue #1: invalid_type at choices
  Required
"
`;

exports[`QuestionSchema missing type 1`] = `
"2 validation issue(s)

  Issue #0: unrecognized_keys at 
  Unrecognized key(s) in object: 'defaultValueFromQuestionId', 'slider'

  Issue #1: invalid_enum_value at type
  Input must be one of these values: Slider, ChoicesWithSingleAnswer, ChoicesWithMultipleAnswers, YesNo, MultipleText, HowLongAgo, Branch, BranchWithRelativeComparison
"
`;

exports[`QuestionSchema next should not be undefined 1`] = `
"1 validation issue(s)

  Issue #0: invalid_type at next
  Required
"
`;

exports[`SliderQuestionSchema default value range 0 - 100: <0 1`] = `
"1 validation issue(s)

  Issue #0: too_small at defaultValue
  Value should be greater than or equal to 0
"
`;

exports[`SliderQuestionSchema default value range 0 - 100: >100 1`] = `
"1 validation issue(s)

  Issue #0: too_big at defaultValue
  Value should be less than or equal to 100
"
`;

exports[`SliderQuestionSchema defaultValueFromQuestionId cannot be null 1`] = `
"1 validation issue(s)

  Issue #0: invalid_union at defaultValueFromQuestionId
  Invalid input
"
`;

exports[`SliderQuestionSchema defaultValueFromQuestionId is question ID 1`] = `
"1 validation issue(s)

  Issue #0: custom_error at defaultValueFromQuestionId
  Question ID can only include letters, numbers, and \\"_\\".
"
`;

exports[`SliderQuestionSchema slider tuple: one item 1`] = `
"1 validation issue(s)

  Issue #0: too_small at slider
  Should have at least 2 items
"
`;

exports[`SliderQuestionSchema slider tuple: three items 1`] = `
"1 validation issue(s)

  Issue #0: too_big at slider
  Should have at most 2 items
"
`;

exports[`SliderQuestionSchema slider tuple: zero item 1`] = `
"1 validation issue(s)

  Issue #0: too_small at slider
  Should have at least 2 items
"
`;

exports[`SliderQuestionSchema type 1`] = `
"1 validation issue(s)

  Issue #0: invalid_literal_value at type
  Input must be \\"Slider\\"
"
`;
